### ISSUE TITLE
Add Pug (formerly Jade) support

### ISSUE THREAD

**Issue by jerone** [Glavin001/atom-beautify#199]
Would be great if atom-beautify supported [Jade](http://jade-lang.com/) template language. :+1: 

:x: `Unsupported language for grammar 'Jade'.`

**Comment by Bodhiz** [comment]
+1

**Comment by scarnago** [comment]
I love this atom plugin, but I need Jade template language support

+1

**Comment by mez** [comment]
I was just going to suggest this!

**Comment by Glavin001** [comment]
@prettydiff is this something Prettydiff could support? [I have not found another yet.](http://stackoverflow.com/questions/4989467/jade-template-engine-html-beautifier)

**Comment by prettydiff** [comment]
No, this one will not be supported by Pretty Diff.  I used Jade before and it is a fantastic idea to write HTML faster, but unfortunately it is white-space and grammar based instead of syntax based.  There is nothing in Pretty Diff to accommodate this language without writing a new parser from scratch.

**Comment by Glavin001** [comment]
Makes sense. I believe there was a “pretty” option in Node.js’ Jade package, I’ll look into that later. Thanks.

**Comment by jerone** [comment]
@Glavin001 commented on [16 mrt. 2015 18:08 CET](https://github.com/Glavin001/atom-beautify/issues/199#issuecomment-81813655 "2015-03-16T17:08:51Z - Replied by Github Comment Enhancer"):
> <div class="email-fragment">Makes sense. I believe there was a “pretty” option in Node.js’ Jade package, I’ll look into that later. Thanks.</div>
> <span class="email-hidden-toggle">[…](#)</span><div class="email-hidden-reply" style="display:none">
> <div class="email-quoted-reply"> On Mar 16, 2015, at 2:07 PM, Austin Cheney &lt;notifications@github.com&gt; wrote:
> 
>  No, this one will not be supported by Pretty Diff. I used Jade before and it is a fantastic idea to write HTML faster, but unfortunately it is white-space and grammar based instead of syntax based. There is nothing in Pretty Diff to accommodate this language without writing a new parser from scratch.
> 
>  —
>  Reply to this email directly or view it on GitHub &lt;[https://github.com/Glavin001/atom-beautify/issues/199#issuecomment-81813117](https://github.com/Glavin001/atom-beautify/issues/199#issuecomment-81813117)&gt;.
> </div>
> <div class="email-fragment"></div>
> </div>

The `pretty` option from Jade is only to prettify the generated HTML (newlines, indentation), not the jade files themselves.

**Comment by josh-egan** [comment]
+1 for adding Jade support :smiley:

**Comment by Evrei** [comment]
+1 plz need Jade support!

**Comment by Glavin001** [comment]
I currently do not know of any beautifiers that handle Jade. If you know of one let me know. Thanks.

**Comment by GuidoS** [comment]
+1 plz

**Comment by Glavin001** [comment]
Could users provide with some small examples of input and expected beautifier outputs? I cannot find a beautifier to support Jade however maybe one could be made for the specific usage that each of you users needs. A lot of posts I am finding about Jade beautification / formatter state simply that since it is a whitespace based language it should implicitly be formatted correctly, by the user. So providing some clear examples to me would help me fine tune exactly what needs to be done. Thanks!

**Comment by pjbrow** [comment]
+1 from me too, and good point @Glavin001 - I've been waiting for someone to say this.  I think what everyone has in mind would do a lot less than a beautifier for JS or HTML.  Here's an example I come across quite often that I'd use it for.

**Input**:  Excess tabs - you might have deleted something in between.
```
div.first-div
		div.first-sub-div
			p First paragraph
		div.second-sub-div
```
**Output**:  Excess tabs gone.
```
div.first-div
	div.first-sub-div
		p First paragraph
	div.second-sub-div
```

**Comment by jerone** [comment]
I can think of the following actions a beautifier could solve (maybe some are more linters):

* As @pjbrow mentions, indention alignment is one that a beautifier could solve.
* As JetBrains [mentions](https://www.jetbrains.com/webstorm/help/code-style-jade.html), tabs to spaces and visa versa.
* Omit the `div` element when it has a [class or id](http://jade-lang.com/reference/attributes/).
* Force/remove comma's `,` between [attributes](http://jade-lang.com/reference/attributes/).
* Force/remove spaces between [attributes](http://jade-lang.com/reference/attributes/).
* For negated [conditions](http://jade-lang.com/reference/conditionals/) use `unless`.
* Use basic pattern for escaped string [interpolation](http://jade-lang.com/reference/interpolation/).
* Use block expansion for simple nested [tags](http://jade-lang.com/reference/tags/).
* Self closing [tags](http://jade-lang.com/reference/tags/).

**Comment by prettydiff** [comment]
@jerone could you provide an example of negated condition syntax, self closing tags?

You likely would not get the 4th and 5th bullet points from a beautifier. Software would not be intelligent enough to know where a code author makes mistakes in opposition to their intentions. At that point the code should be allowed to be broken and the code author should test it and fix it accordingly. A beautifier attempting to make this decision will always be wrong at some point.

**Comment by myrtleTree33** [comment]
+1

**Comment by vingorius** [comment]
+1

**Comment by vingorius** [comment]
Today i developed a simple program for jade beautify.

I'm a newbie , code is dirty.

It's main function is only reformat it's indent.
(check upper @pjbrow 's comment)

to install, search 'jade beautify' on package install.

open jade file which you want to beautify, click 'ctrl-alt-o'.

enjoy jade.

**Comment by vingorius** [comment]
# Jade Beautify
I made a tiny 'jade beautify' for myself.
Everyone can install it, just search 'jade beautify'.
and welcome feedback.

Source: https://github.com/vingorius/jade-beautify

Enjoy jade.

**Comment by Glavin001** [comment]
@vingorius: Awesome! Would you be interested in:
- creating / publishing a modular Node.js package, such that jade-beautify could be a reusable package for Node.js that others could use and contribute to.
- submitting a Pull Request to Atom Beautify to add your Jade Beautify as the Jade beautifier. See docs: https://github.com/Glavin001/atom-beautify/blob/master/docs/add-languages-and-beautifiers.md#how-to-add-a-beautifier-for-a-language

Great work! Thanks.

**Comment by vingorius** [comment]
@Glavin001  Thanks for your suggest. I'll check.

**Comment by ryanpcmcquen** [comment]
+1, @vingorius let me know if I can help in any way!

**Comment by vingorius** [comment]
Thanks @ryanpcmcquen . 
I had checked atom-beautify. it was too huge to understand as a beginner of coffeescript.

If you can help me, it's very appreciated.

**Comment by jaseeey** [comment]
+1, I would love to see this integrated into this package, then it would be complete for me!

**Comment by pauljakobwhite** [comment]
+1

**Comment by danielbayley** [comment]
+1.

**Comment by ThorstenHans** [comment]
keep me posted (or short +1)

**Comment by jonreader** [comment]
+1

**Comment by Glavin001** [comment]
If someone would publish to npm a `jade beautify` package that Atom Beautify could use then I could quickly add support for this by following the instructions found at https://github.com/Glavin001/atom-beautify/blob/master/docs/add-languages-and-beautifiers.md#how-to-add-a-beautifier-for-a-language .
/cc @vingorius

**Comment by Glavin001** [comment]
Jade is changing its name to Pug: https://github.com/pugjs/jade/issues/2184

**Comment by ErikBoesen** [comment]
+1

**Comment by Glavin001** [comment]
I will be adding Jade support this weekend, thanks to @vingorius! If someone would like to create a Pull Request adding support, the package can be found at https://github.com/vingorius/pug-beautify and let me know if you need any help. Thanks!

**Comment by cgdeboer** [comment]
@Glavin001 is there an update on this. I would be a huge fan of `atom-beautify` supporting pug(jade).

**Comment by Glavin001** [comment]
I haven't had a chance to finish adding Jade support. However, I have moved this back to v0.29.0 milestone, so I want it to be in the next release. :+1:

**Comment by Glavin001** [comment]
Published to v0.28.28

**Comment by cgdeboer** [comment]
Thanks. This is great !
